### PR TITLE
Add support for XT300/XH300 soil moisture sensor

### PR DESCRIPTION
**Pull request by enavarro222**
I spent few time on this sensor that I have on a shelf for a while (much simpler than the TPMS ...)

Test data are here : https://github.com/merbanan/rtl_433_tests/pull/258

### CODE REVIEW

File: src/devices/xt300.c [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        temp     = b[3] - 40;
+        temp_c    = temp * 1.;
+        moisture =  b[2];
+
+        data = data_make(
+            "model",            "",             DATA_STRING, "XT300 Temperature & Moisture",
+            "channel",          "Channel",      DATA_INT,    channel,
+            "temperature_C",    "Temperature",  DATA_FORMAT, "%.01f C", DATA_DOUBLE, temp_c,
+            "moisture",         "Moisture",     DATA_FORMAT, "%d %%", DATA_INT, moisture,
+            "mic",              "Integrity",    DATA_STRING, "CHECKSUM",
```
Comment (Mindavi) [inline src/devices/xt300.c]: why is there a field that's hardcoded to CHECKSUM?

Comment (zuckschwerdt) [reply src/devices/xt300.c]: That's a marker to show there is a valid checksum

**Comment by zuckschwerdt** [comment]
Is the manufacturer Imagintronix (or Opus, or are they both resellers)?
The model name should be Imagintronix-Xx300, the filename imagintronix-xx300.c or maybe Soil-Xx300 / soil-xx300.c if the manufacturer isn't clear. Or perhaps …-XT300 if the "XH" isn't too popular (or is XH300 the monitor? then definitely XT300)

File: src/devices/xt300.c (hunk 1 above)
Comment (Mindavi) [reply src/devices/xt300.c]: ah okay, that's useful

**Comment by Mindavi** [comment]
Opus seems like the factory that produces those things. They seem to have a store on alibaba selling these units. The source code of the plantcaretools website also talks about opus xt300.

File: src/devices/opus_xt300.c [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            continue;
+        }
+
+        channel  = (b[1] & 0x03);
+        temp     = b[3] - 40;
+        temp_c    = temp * 1.;
+        moisture =  b[2];
+
+        data = data_make(
+            "model",            "",             DATA_STRING, "Opus/Imagintronix XT300 Soil Moisture",
```
Comment (zuckschwerdt) [inline src/devices/opus_xt300.c]: Can you change the "model" to "Opus-XT300"? The r_device name should stay "Opus/Imagintronix XT300 Soil Moisture".

File: src/devices/opus_xt300.c [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        ret++;
+    }
+    return ret;
+}
+
+static char *output_fields[] = {
+    "model",
+    "sid",
+    "channel",
+    "battery",
```
Comment (zuckschwerdt) [inline src/devices/opus_xt300.c]: there is no battery for now.

File: src/devices/opus_xt300.c [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            NULL);
+        decoder_output_data(decoder, data);
+        ret++;
+    }
+    return ret;
+}
+
+static char *output_fields[] = {
+    "model",
+    "sid",
```
Comment (zuckschwerdt) [inline src/devices/opus_xt300.c]: change to "id"

File: src/devices/opus_xt300.c (hunk 2 above)
Comment (zuckschwerdt) [reply src/devices/opus_xt300.c]: Please add add "id" after the model.

File: src/devices/opus_xt300.c [hunk 5]
```diff
@@ -1,4 +1,6 @@
-/* XH300 / XT300 Soil Moisture Sensor
+/* Opus/Imagintronix XT300 Soil Moisture Sensor
+ *
+ * Aslo call XH300 some time, this seems to be the associated display name
```
[the remark this answers is not in the document]
Comment (enavarro222) [reply src/devices/opus_xt300.c]: yep sorry for my English ;)

File: src/devices/opus_xt300.c [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        }
+        chk = add_bytes(b + 1, 4); // sum bytes 1-4
+        chk = chk & 0xFF;
+        if (chk != 0 & chk != b[5] ) {
+            continue;
+        }
+
+        channel  = (b[1] & 0x03);
+        temp     = b[3] - 40;
+        temp_c    = temp * 1.;
```
[the remark this answers is not in the document]
Comment (enavarro222) [reply src/devices/opus_xt300.c]: right temp is actuality an int

**Comment by zuckschwerdt** [comment]
Looks really good! Preferably `src/CMakeLists.txt` and `src/Makefile.am` should be alphabetically sorted (unlike `include/rtl_433_devices.h`). But I can fix that later if needed.
Can you explain what should be in b[1]? If I read it right `b[1]&0xfc==0x54`, i.e. the two lower bits (`b[1]&3`) are variable, some kind of ID or repeat counter?

**Comment by Net-time** [comment]
Hi. Just bought the RTL-SDR so bit new at this.
Bought it partly because this filter is in the making as I have a Imaginetronix IM300, but I think I wrongly assumed it to be the same as yours as mine looks exactly the same.
I have not had the time to check how to produce test data as yet, but I have previosly decoded my sensor and the result was:

 Imaginetronix IM 300. 28 bits, long bit 0, 2-3 repetions
 __________________Ch..........% 7b________°Clsb ___________ °Cmsb
11111110 1010 1010 0000101 0 1000 011 1 11111111 0011 1000    
11111110 1010 1100 0000101 0 1000 011 1 11111111 0011 1010    
11111110 1010 1110 0000101 0 1000 100 1 11111111 0011 1110    
seems I forgot to test how negative temperature are indicated.
But with a quick look at the XT-300 filter they dont seem to match.
I'm not fluent in programming, so If you could confirm that they are not the same I'll start a new Pull Request.

**Comment by zuckschwerdt** [comment]
@Net-time The only thing open here IIRC is to label the first 2 bits. The layout exactly fits your description. Can you explain what is in "Ch" is it counting, is it set randomly on replacing the battery or is it indeed a selectable channel setting? Looks like a repeat-counter in your example?

**Comment by zuckschwerdt** [comment]
Here is a [BitBench](http://triq.net/bitbench?c=y%2011111110%201010%201010%200000101%200%201000%20011%201%2011111111%200011%201000&c=y%2011111110%201010%201100%200000101%200%201000%20011%201%2011111111%200011%201010&c=y%2011111110%201010%201110%200000101%200%201000%20100%201%2011111111%200011%201110&f=PRE%3A7b%20FIX%3A6b%20CH%3A2b%20SOIL%3A8d%20TEMP%3A8d%20%3F%3A8b%20CHK%3A8h%201x) to better show the data layout. (Substract 40 from temp, the example is 27 to 28).

**Comment by Net-time** [comment]
Ohh, that's good news :)
They look as a counter as It's the diffrent channel codes, first row ch1, second ch2 etc. max 3 channels.
You select channel by moving a single Jumper in the sensor.
Guessing I never discovered the "-40" part as it's not needed if you only test positive numbers.

**Comment by zuckschwerdt** [comment]
Great. I'll merge this PR in a few days and then your sensors should work with stock rtl_433.

**Comment by enavarro222** [comment]
Hi, thanks for the up on this. I had very few times to finalize this PR.
And yes : b[1]&3 is the channel number.
No random id or so.

**Comment by Net-time** [comment]
Thanks, works fine.
I checked if there is a battery indicator in the raw data, it does not have one.
Do not show moisture in Domoticz as it's not included in the parse in Domoticz/hardware/Rtl433.cpp so I opened a Issue there.

**Comment by geekdave** [comment]
I was super excited to find this thread but I can't seem to find anywhere online that sells the sensors listed above.  I've tried alibaba and Imagintronix websites.  

Anyone have a current link for a soil moisture sensor that's compatible with the above PR?  (I'm in the USA)

Thanks!  We're building a garden and I'd love to try this out!

**Comment by Net-time** [comment]
Hi!
https://www.ebay.ie/itm/100m-Wireless-Transmission-Soil-Humidity-Tester-Multifunction-Indoor-Outdoor-Tem/162453974127?hash=item25d302ec6f:m:m6gdsfhg02R8EXhdX77bh1g
You can only have 3 sensors so if you want more you should look for another hardware.
Also they measure air temperature, not soil temperature.
Last time i checked Domoticz does not parse the sensor correctly so I recompiled Rtl433 to send humidity instead of moisture which works.

**Pull request merged**